### PR TITLE
Batch U projection all-reduces into single NCCL collective

### LINKED ISSUE #71: Batch U projection all-reduces into single NCCL collective
### Problem

Each OSFT target's U projection requires an all-reduce of a `(k, k_low)` coefficient matrix:

```python
proj_coeff = torch.mm(local_U_high.transpose(0, 1), local_dU)
dist.all_reduce(proj_coeff, op=dist.ReduceOp.SUM)
```

For Llama-8B with 224 targets, this is **224 separate `dist.all_reduce` calls** per optimizer step.  Each call incurs NCCL kernel launch overhead, synchronization, and GPU scheduling — measured at **~56 μs per call** on 2× H200 NVLink.  The individual payloads are small (~4M elements for a 4096-dim layer at URR=0.5), making this **latency-dominated** rather than bandwidth-dominated.

### Proposed solution

Restructure `project_gradients` into three phases for the distributed case:

**Phase 1 — Compute:** Loop over all OSFT modules, compute local U projection coefficients `U_high^T @ dU` via matmul (no all-reduce). Flatten and collect each coefficient tensor.

**Phase 2 — Batch all-reduce:** `torch.cat` all flattened coefficients into one tensor, single `dist.all_reduce`.

**Phase 3 — Apply:** Offset-split the all-reduced tensor back into per-target coefficients and apply each projection.

V projections reuse the existing `project_gradient_to_orthogonal_space(svd_dict, skip_u=True)` — no code duplication.  Single-GPU path falls through unchanged.

### Why batching is valid

All-reduce(SUM) distributes over concatenation: `allreduce(cat([a, b])) = cat([allreduce(a), allreduce(b)])`.  The per-target coefficient matrices are independent, so the batched result is **mathematically identical** to the unbatched path.

### GPU memory implications

The batched tensor is a single contiguous allocation holding all flattened U projection coefficients:

| Model | Targets | Avg coeff size | Batched tensor | Overhead |
|---|---|---|---|---|
| **Llama-8B** | 224 | (k, k_low) ≈ 4M elements | ~1.4 GB (bf16) | Temporary, freed after Phase 3 |
| **Llama-70B** | 560 | ~16M elements | ~14 GB (bf16) | Temporary |

This is a **transient** allocation — it exists only during the `project_gradients` call and is freed immediately after.  No persistent memory overhead.

### Impact

| Metric | Before | After |
|---|---|---|
| U all-reduce kernel launches per step | 224 | 1 |
| U all-reduce data volume per step | ~1.4 GB | ~1.4 GB (unchanged) |
| U all-reduce wall time (2× H200 NVLink) | 12.4 ms | 7.1 ms (**1.8× faster**) |

### Mathematical justification

Ref to your nice paper -  Eq. 4 from Nayak et al., "Sculpting Subspaces" ([arXiv:2504.07097](https://arxiv.org/abs/2504.07097), Theorem 1).

### PR DESCRIPTION
**Pull request by stmcgovern**
Closes #71

Batch all per-target U projection coefficient all-reduces into a single
`dist.all_reduce` call. Reduces 224 NCCL kernel launches to 1 per step
(Llama-8B). Data volume unchanged — this targets launch latency, not bandwidth.

**Benchmark (2× H200, NVLink, 224 targets):**

| | Individual | Batched |
|---|---|---|
| **Wall time** | 12.4 ms | 7.1 ms |
| **Speedup** | — | **1.8×** |
| **NCCL launches** | 224 | 1 |
| **Per-call overhead** | 56 μs | — |

Single-GPU path unchanged. Transient memory only (~1.4 GB for Llama-8B,
freed after `project_gradients` returns).

### Changes

- `project_gradient_to_orthogonal_space` gains `skip_u` kwarg (default `False`)
- `project_gradients` restructured into compute → batch all-reduce → apply
  for the distributed path
- 5 new tests including `test_batched_path_matches_unbatched` (bitwise equality
  vs unbatched path)

## Release Notes

* **New Features**
  * Added optional parameter to control gradient projection behavior in distributed training scenarios.
  * Implemented batched all-reduce optimization for improved distributed training performance across multiple modules.

* **Tests**
  * Added comprehensive test coverage for batched gradient projection behavior and multi-module optimization scenarios.

### CODE REVIEW

**Review by RobotSail — APPROVED** [review]
Tested and it seems to work as expected, I am seeing a training speedup of around 2%. Thanks for this change, LGTM!
